### ISSUE TITLE
Custom JSONLoader

### ISSUE THREAD

**Issue by yauhen-l** [xeipuuv/gojsonschema#99]
Why all methods in JSONLoader interface are private?
https://github.com/xeipuuv/gojsonschema/blob/master/jsonLoader.go#L45
As result there is no way to implement customer JSONLoader.

My use case:
I use go-bindata (https://github.com/jteeuwen/go-bindata) to inject JSON schemas into Go binary.
To validate JSON messages I use `gojsconschema.Validate(gojsonschema.NewStringLoader(Asset("SCHEMA_KEY")), gojsonschema.NewStringLoader(message))`.
But I also would like to have an opportunity to reference to another schema in assets.
For example, if I have in JSON schema:
`...
   "$ref": "/common/some-object"
...
`
Look will be in Asset("/common/some-object").

In other words references pool should be filled up with result of Asset() function.

**Comment by cristiangraz** [comment]
I agree, I was going to implement a custom `JSONLoader` from a `Reader` a while back so there isn't a need to go from `Reader` to string, then have the library convert the string back to a `Reader`. Ran into this same issue preventing me from doing that.

@xeipuuv Would you be opposed to have the interface methods be public? Given that `JSONLoader` is public, it seems this may have been the intention originally? If @yauhen-l is busy, I can work on this over the next few weeks.

**Comment by xeipuuv** [comment]
As @cristiangraz pointed out, JSONLoader is public, and yes the intention was to allow for custom loaders to be implemented; nobody asked for custom loaders yet so i guess the private functions issue remained.

@yauhen-l You can't really reference an external schema when using the string loader ( they are considered standalone ). You could write your own loader, but i am not sure this is enough and if it is even possible to do that in an elegant way. To do what you want i would use a file loader instead.

@cristiangraz No problems with making the interface methods public.

**Comment by yauhen-l** [comment]
@xeipuuv not sure how file loader could help me.
Actually `Asset("something")` is an alias of map access:
`
func Asset(key string) ([]byte, error) {
...
  return values[key]
} 
`

**Comment by xeipuuv** [comment]
File loaders, and reference loaders allows you to use external schema references. String loaders dont.
Of course in order to use file loaders you need files instead of the Asset thing.

**Comment by yuichi1004** [comment]
We also have an issue implementing our custom JSON loader.

Our use case:
 - We want to validate request payload on our webapp
 - We want to pusblish our JSON schema using 'https://' scheme so that 3rd party can easily resues the JSON schema to implement their client
 - While using 'https://' schema, we would like to load our JSON schema from local file for the following reason
    - Performance on webapp server
    - Edit the schema locally for the development

We have tried to use #101 to implement the above but did not work because we use 'https://' reference notation.

I like JSONLoader to be public interface so that we can flexibly customize our JSON load process as you guys already discussed the above. I will work on this issue if you have difference opinion.
